src/protocol/range_spec.rs

```diff
@@ -0,0 +1,348 @@
+        }
+    }
+}
+
+/// A compressed sequence of range specs
+///
+/// Examples:
+///
+/// All child ranges: `[(0, [0])]` starting at offset 0, all offsets (see above)
+/// First chunk of all children: `[(0, [0, 1])]` starting at offset 0, chunk range 0..1
+/// All of child 1234: `[(1234, [0]), [1, []]]`.
+/// First 33 chunks of child 5678: `[(5678, [0, 33]), (1, [])]`.
+/// Chunks 10 to 30 of child 6789: `[(6789, [10, 20]), (1, [])]`.
+/// No child ranges: `[]`
+///
+/// This is a smallvec so that we can avoid allocations in the common case of a single child range.
+#[derive(Deserialize, Serialize, Debug, PartialEq, Eq, Clone)]
+#[repr(transparent)]
+pub struct RangeSpecSeq(SmallVec<[(u64, RangeSpec); 2]>);
+
+impl RangeSpecSeq {
+    #[allow(dead_code)]
+    /// An empty range spec sequence
+    ///
+    /// When iterated, will return an empty range forever
+    pub const fn empty() -> Self {
+        Self(SmallVec::new_const())
+    }
+
+    /// If this range seq describes a range for a single item, returns the offset
+    /// and range spec for that item
+    pub fn single(&self) -> Option<(u64, &RangeSpec)> {
+        // we got two elements,
+        // the first element starts at offset 0,
+        // and the second element is empty
+        if self.0.len() != 2 {
+            return None;
+        }
+        let (fst_ofs, fst_val) = &self.0[0];
+        let (snd_ofs, snd_val) = &self.0[1];
+        if *fst_ofs == 0 && *snd_ofs == 1 && snd_val.is_empty() {
+            Some((*fst_ofs, fst_val))
+        } else {
+            None
+        }
+    }
+
+    /// A complete range spec sequence
+    ///
+    /// When iterated, will return a full range forever
+    pub fn all() -> Self {
+        Self(smallvec![(0, RangeSpec::all())])
+    }
+
+    /// Create a new range spec sequence from a sequence of range sets
+    pub fn new(children: impl IntoIterator<Item = RangeSet2<ChunkNum>>) -> Self {
+        let mut prev = RangeSet2::empty();
+        let mut count = 0;
+        let mut res = SmallVec::new();
+        for v in children
+            .into_iter()
+            .chain(std::iter::once(RangeSet2::empty()))
+        {
+            if v == prev {
+                count += 1;
+            } else {
+                res.push((count, RangeSpec::new(&v)));
+                prev = v;
+                count = 1;
+            }
+        }
+        Self(res)
+    }
+
+    /// An infinite iterator of range specs
+    pub fn iter(&self) -> RequestRangeSpecIter<'_> {
+        let before_first = self.0.get(0).map(|(c, _)| *c).unwrap_or_default();
+        RequestRangeSpecIter {
+            current: &EMPTY_RANGE_SPEC,
+            count: before_first,
+            remaining: &self.0,
+        }
+    }
+
+    /// An iterator over non empty range specs
+    ///
+    /// This iterator is infinite if the range spec is infinite
+    pub fn non_empty_iter(&self) -> NonEmptyRequestRangeSpecIter<'_> {
```

Comment:
bikeshedding: `iter_non_empty`.  this maches e.g. `iter_mut` from the stdlib where the specialisation is appended to the core thing.